### PR TITLE
[BIOMAGE-2260] Refactor nPCs estimation in the integration step

### PR DESCRIPTION
**Pull request by saracastel**
# Description
This PR refactors the way in which we handle nPCs in every integration method. When we run an integration method different from harmony, the elbow plot doesn't show the full number of principal components and the nPCs can't be increased because the % of variance explained is wrongly set to 100%. This PR fixes that by running PCA using the max number of PCs (50 or the number of cells of the smallest sample if smaller than 50).
It also fixes some things that were causing geosketch to fail.
Lastly, it removes the old integration code.

# Details
#### URL to issue
https://biomage.atlassian.net/browse/BIOMAGE-2260?atlOrigin=eyJpIjoiODNlYTcxOGU0MDQ5NDUzYTk1MmE5YjI1YmYzZmUwNjIiLCJwIjoiaiJ9

#### Link to staging deployment URL (or set N/A)
https://ui-sara-pipeline136.scp-staging.biomage.net

#### Integration test branch
master

# Merge checklist
Your changes will be ready for merging after all of the steps below have been completed.

### Code updates
Have best practices and ongoing refactors being observed in this PR
- [ ] Migrated any selector / reducer used to the new format.

### Manual/unit testing
- [ ] Tested changes using InfraMock locally **or** no tests required for change, e.g. Kubernetes chart updates.
- [ ] Validated that current unit tests for code work as expected and are sufficient for code coverage **or** no unit tests required for change, e.g. documentation update.
- [ ] Unit tests written **or** no unit tests required for change, e.g. documentation update.

### Integration testing
**You must check the box below** to run integration tests on the latest commit on your PR branch.
Integration tests have to pass before the PR can be merged. Without checking the box, your PR
**will not pass** the required status checks for merging.

- [x] Started end-to-end tests on the latest commit.

### Documentation updates
- [ ] Relevant Github READMEs updated **or** no GitHub README updates required.
- [ ] Relevant Wiki pages created/updated **or** no Wiki updates required.

### Optional
- [ ] Staging environment is unstaged before merging.
- [ ] Photo of a cute animal attached to this PR.

### CODE REVIEW

File: pipeline-runner/R/qc-6-integrate_scdata-fastmnn.R [hunk 1]
```diff
@@ -24,29 +24,18 @@ run_fastmnn <- function(scdata_list, config, cells_id) {
...
-  scdata@assays$RNA@scale.data <- as.matrix(0)
 
   if (use_geosketch) {
     scdata <-
       RunGeosketchFastMNN(
         scdata,
         split.by = "samples",
         features = nfeatures,
+        npcs_for_pca,
         dims = npcs,
```
Comment (gerbeldo) [inline pipeline-runner/R/qc-6-integrate_scdata-fastmnn.R:39]: is this OK?

Comment (gerbeldo) [reply pipeline-runner/R/qc-6-integrate_scdata-fastmnn.R:39]: oh, I see you added the argument in the RunGeosketchFastMNN function, you should name it for consistency

File: pipeline-runner/R/qc-6-integrate_scdata-fastmnn.R [hunk 2]
```diff
@@ -85,23 +74,31 @@ prepare_scdata_for_fastmnn <- function(scdata_list, config, cells_id) {
...
     perc_num_cells = perc_num_cells
-    )
+  )
 
   scdata_split <- Seurat::SplitObject(geosketch_list$sketch, split.by = split.by)
   scdata_sketch_integrated <-
     SeuratWrappers::RunFastMNN(
       scdata_split,
-      features = nfeatures,
+      features = features,
```
Comment (gerbeldo) [inline pipeline-runner/R/qc-6-integrate_scdata-fastmnn.R:97]: good catch!

File: pipeline-runner/R/qc-6-integrate_scdata-harmony.R [hunk 3]
```diff
@@ -28,7 +27,13 @@ run_harmony <- function(scdata_list, config, cells_id) {
...
     Seurat::FindVariableFeatures(nfeatures = nfeatures, verbose = FALSE) |>
     Seurat::ScaleData(verbose = FALSE) |>
-    Seurat::RunPCA(npcs = npcs_for_pca, verbose = FALSE)
+    # run PCA with npcs_for_pca for the elbow plot and the % of variance explained
+    Seurat::RunPCA(
+      npcs = npcs_for_pca,
+      reduction.name = "pca_for_plot",
+      reduction.key = "PCp_",
+      verbose = FALSE
+    )
```
Comment (gerbeldo) [inline pipeline-runner/R/qc-6-integrate_scdata-harmony.R:36]: given that, except for harmony, it does not make sense to run PCA twice, I would leave the reduction names and keys to the default values. (except the workaround for harmony)

File: pipeline-runner/R/qc-6-integrate_scdata-harmony.R [hunk 4]
```diff
@@ -96,25 +101,41 @@ prepare_scdata_for_harmony <- function(scdata_list, config, cells_id) {
 RunGeosketchHarmony <- function(scdata, group.by.vars, reduction, dims.use, config) {
   set.seed(RANDOM_SEED)
   perc_num_cells <- config$downsampling$methodSettings$geosketch$percentageToKeep
+  dims <- dims.use[length(dims.use)]
```
Comment (gerbeldo) [inline pipeline-runner/R/qc-6-integrate_scdata-harmony.R:104]: when calling `RunGeosketchHarmony` we could pass `npcs` directly, instead of `1:npcs`, avoiding adding this line.

File: pipeline-runner/R/qc-6-integrate_scdata-harmony.R [hunk 5]
```diff
@@ -96,25 +101,41 @@ prepare_scdata_for_harmony <- function(scdata_list, config, cells_id) {
...
-      dims.use = dims.use
-    )
+  if (ncol(geosketch_list$sketch) < 15) {
+    scdata_sketch_integrated <-
+      harmony::RunHarmony(
+        geosketch_list$sketch,
+        group.by.vars = "samples",
+        reduction = "pca_for_harmony",
+        dims.use = dims.use,
+        nclust = ncol(geosketch_list$sketch) - 1
```
Comment (gerbeldo) [inline pipeline-runner/R/qc-6-integrate_scdata-harmony.R:119]: why is this necessary? does it make sense to downsample to a total of less than 15 cells? shouldn't we instead limit the minimum number of cells when geosketching rather than doing something like this?

File: pipeline-runner/R/qc-6-integrate_scdata-harmony.R [hunk 6]
```diff
@@ -96,25 +101,41 @@ prepare_scdata_for_harmony <- function(scdata_list, config, cells_id) {
...
+  }
+
+  scdata_sketch_integrated@misc[["active.reduction"]] <- "harmony"
 
   scdata <- learn_from_sketches(
     geosketch_list$scdata,
     geosketch_list$sketch,
     scdata_sketch_integrated,
-    dims = dims.use
+    dims = dims,
```
Comment (gerbeldo) [inline pipeline-runner/R/qc-6-integrate_scdata-harmony.R:137]: same as before, pass the number straight and use the original argument.

File: pipeline-runner/R/qc-6-integrate_scdata-seuratv4.R [hunk 7]
```diff
@@ -27,9 +26,13 @@ run_seuratv4 <- function(scdata_list, config, cells_id) {
...
   use_geosketch <- "downsampling" %in% names(config) && config$downsampling$method == "geosketch"
 
+  # calculate as many PCs for the PCA as possible, ideally 50, unless few cells
+  npcs_for_pca <- min(vapply(scdata_list, ncol, integer(1)) - 1, 50)
   npcs <- config$dimensionalityReduction$numPCs
+
+  # use the min of what the user wants and what can be calculated
   if (is.null(npcs)) {
-    npcs <- 30
+    npcs <- min(config$dimensionalityReduction$numPCs, npcs_for_pca)
```
Comment (gerbeldo) [inline pipeline-runner/R/qc-6-integrate_scdata-seuratv4.R:35]: this will never run here, because npcs will never be null when runnning seurat (but already set when running harmony. This line would be most useful when setting the npcs, 3 lines above.

File: pipeline-runner/R/qc-6-integrate_scdata-seuratv4.R [hunk 8]
```diff
@@ -180,7 +184,17 @@ seuratv4_find_and_integrate_anchors <-
...
-    # run PCA
+    # run PCA with 50 PCs (or less if there are less cells) for the elbow plot and the % of variance explained
+    scdata <-
+      Seurat::RunPCA(
+        scdata,
+        npcs = npcs_for_pca,
+        reduction.name = "pca_for_plot",
+        reduction.key = "PCp_",
+        verbose = FALSE
+      )
```
Comment (gerbeldo) [inline pipeline-runner/R/qc-6-integrate_scdata-seuratv4.R:195]: remove this, but pass npcs_for_pca to the RunPCA below (because we want those 50 PCs, and we don't need to create this new "pca_for_plot" reduction.

File: pipeline-runner/R/qc-6-integrate_scdata-seuratv4.R [hunk 9]
```diff
@@ -221,6 +235,7 @@ integrate_using_geosketch <-
            perc_num_cells,
            normalization,
            npcs,
+           npcs_for_pca,
            nfeatures,
            use_geosketch) {
```
Comment (gerbeldo) [inline pipeline-runner/R/qc-6-integrate_scdata-seuratv4.R:240]: this function should live in the subsample geosketch file. Or change the name to something more specific if it's only for seurat.

File: pipeline-runner/R/qc-6-integrate_scdata-seuratv4.R [hunk 10]
```diff
@@ -256,6 +271,16 @@ integrate_using_geosketch <-
...
     )
 
+    # add back the pca_for_plot slot which is lost after learn_from_sketches (required for the elbow plot)
+    scdata@reductions[["pca_for_plot"]] <- Seurat::CreateDimReducObject(
+      embeddings = scdata_sketch_integrated@reductions$pca_for_plot@cell.embeddings,
+      loadings = scdata_sketch_integrated@reductions$pca_for_plot@feature.loadings,
+      projected = scdata_sketch_integrated@reductions$pca_for_plot@feature.loadings.projected,
+      stdev = scdata_sketch_integrated@reductions$pca_for_plot@stdev,
+      key = "PCp_",
+      misc = scdata_sketch_integrated@reductions$pca_for_plot@misc
```
Comment (gerbeldo) [inline pipeline-runner/R/qc-6-integrate_scdata-seuratv4.R:281]: same thing about the PCA key

File: pipeline-runner/R/qc-6-integrate_scdata-unisample.R [hunk 11]
```diff
@@ -28,11 +28,21 @@ run_unisample <- function(scdata_list, config, cells_id) {
...
 
+  # run PCA with 50 PCs (or less if there are less cells) for the elbow plot and the % of variance explained
+  scdata <-
+    Seurat::RunPCA(
+      scdata,
+      npcs = npcs_for_pca,
+      reduction.name = "pca_for_plot",
+      reduction.key = "PCp_",
+      verbose = FALSE
+    )
```
Comment (gerbeldo) [inline pipeline-runner/R/qc-6-integrate_scdata-unisample.R:44]: same comment, remove and keys.

File: pipeline-runner/R/qc-6-integrate_scdata.R [hunk 12]
```diff
@@ -147,7 +147,7 @@ get_explained_variance <- function(scdata) {
   if (scdata@misc[["active.reduction"]] == "mnn") {
     var_explained <- scdata@tools$`SeuratWrappers::RunFastMNN`$pca.info$var.explained
   } else {
-    eig_values <- (scdata@reductions$pca@stdev)^2
+    eig_values <- (scdata@reductions$pca_for_plot@stdev)^2
```
Comment (gerbeldo) [inline pipeline-runner/R/qc-6-integrate_scdata.R:150]: key

File: pipeline-runner/R/qc-6-integrate_scdata-fastmnn.R (hunk 1 above)
Comment (saracastel) [reply pipeline-runner/R/qc-6-integrate_scdata-fastmnn.R]: renamed

File: pipeline-runner/R/qc-6-integrate_scdata-harmony.R (hunk 5 above)
Comment (saracastel) [reply pipeline-runner/R/qc-6-integrate_scdata-harmony.R:116]: yeah I was not sure about this. I ended up doing this because my reasoning was that usually, for similar cases, we take an approach that allows the user to do it even if it makes little sense. Then it's up to the user to decide wether or not to trust the result. But I'm willing to change it if you think it would be better to limit the min number of cells

File: pipeline-runner/R/qc-6-integrate_scdata-seuratv4.R (hunk 7 above)
Comment (saracastel) [reply pipeline-runner/R/qc-6-integrate_scdata-seuratv4.R:35]: right!

File: pipeline-runner/R/qc-6-integrate_scdata-seuratv4.R (hunk 9 above)
Comment (saracastel) [reply pipeline-runner/R/qc-6-integrate_scdata-seuratv4.R:224]: renamed to `seuratv4_geosketch_find_and_integrate_anchors` because it's only for seurat

File: pipeline-runner/R/qc-6-integrate_scdata-harmony.R [hunk 13]
```diff
@@ -45,7 +48,7 @@ run_harmony <- function(scdata_list, config, cells_id) {
       scdata,
       npcs = npcs,
       verbose = FALSE,
-      reduction.name = "pca_for_harmony",
+      reduction.name = "pca",
```
Comment (gerbeldo) [inline pipeline-runner/R/qc-6-integrate_scdata-harmony.R:51]: reduction.name here should be "pca_for_harmony". this is the workaround.

Comment (gerbeldo) [inline pipeline-runner/R/qc-6-integrate_scdata-harmony.R:51]:
```suggestion
      reduction.name = "pca_for_harmony",
```

File: pipeline-runner/R/qc-6-integrate_scdata-harmony.R [hunk 14]
```diff
@@ -56,16 +59,16 @@ run_harmony <- function(scdata_list, config, cells_id) {
       RunGeosketchHarmony(
         scdata,
         group.by.vars = "samples",
-        reduction = "pca_for_harmony",
-        dims.use = 1:npcs,
+        reduction = "pca",
```
Comment (gerbeldo) [inline pipeline-runner/R/qc-6-integrate_scdata-harmony.R:62]:
```suggestion
        reduction = "pca_for_harmony",
```

File: pipeline-runner/R/qc-6-integrate_scdata-harmony.R (hunk 5 above)
Comment (gerbeldo) [reply pipeline-runner/R/qc-6-integrate_scdata-harmony.R:116]: yeah, definitely limit the number of cells! downsampling on a 10k dataset (which IMHO does not really make sense) to the 5% default gives you 500 cells. I would even use something larger than that as a minimum, like 1000 or 2000 cells.

File: pipeline-runner/R/qc-6-integrate_scdata-harmony.R (hunk 13 above)
Comment (saracastel) [reply pipeline-runner/R/qc-6-integrate_scdata-harmony.R]: yeah this was not intended. Not sure what happened. Changed it
